sdk/inc/azure/core/az_mqtt.h

```diff
@@ -0,0 +1,391 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+/**
+ * @file
+ *
+ * @brief This header defines the types and functions your application uses to leverage MQTT pub/sub
+ * functionality.
+ *
+ * @details For more details on Azure IoT MQTT requirements please see
+ * https://docs.microsoft.com/azure/iot-hub/iot-hub-mqtt-support.
+ * - Outbound APIs will be called by the SDK to send data over the network.
+ * - Inbound API calls must be called by the MQTT implementation to send data towards the SDK.
+ *
+ * @note Object lifetime: all APIs have run-to-completion semantics. Data passed into the APIs
+ *       is owned by the API for the duration of the call.
+ *
+ * @note API I/O model: It is generally expected that all operations are asynchronous. An outbound
+ *       call should not block, and may occur on the same call-stack, as a result of an inbound
+ *       call.
+ *
+ * @note The SDK expects that the network stack is stalled for the duration of the API calls.
+ *
+ * @note All API implementations must return az_result.
+ * We recommend using `_az_RESULT_MAKE_ERROR(_az_FACILITY_IOT_MQTT, mqtt_stack_ret);` if compatible.
+ *
+ * @note You MUST NOT use any symbols (macros, functions, structures, enums, etc.)
+ * prefixed with an underscore ('_') directly in your application code. These symbols
+ * are part of Azure SDK's internal implementation; we do not document these symbols
+ * and they are subject to change in future versions of the SDK which would break your code.
+ */
+
+#ifndef _az_MQTT_H
+#define _az_MQTT_H
+
+#include <azure/core/az_config.h>
+#include <azure/core/az_context.h>
+#include <azure/core/az_event.h>
+#include <azure/core/az_log.h>
+#include <azure/core/az_result.h>
+#include <azure/core/az_span.h>
+#include <stdbool.h>
+#include <stdint.h>
+
+#if defined(TRANSPORT_MOSQUITTO)
+#include <azure/platform/az_mqtt_mosquitto.h>
+#else
+#include <azure/platform/az_mqtt_notransport.h>
+#endif
+
+#include <azure/core/_az_cfg_prefix.h>
+
+/**
+ * @brief Defines the key storage types.
+ */
+typedef enum
+{
+  AZ_MQTT_X509_CLIENT_CERTIFICATE_KEY_MEMORY = 0,
```

Comment:
Throughout the file: please add doxygen documentation to all enum and struct fields.